crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -4935,22 +5843,357 @@ impl CircuitHandle {
+
+            // self.circuit.to_dot_file(
+            //     |node| {
+            //         if !node.global_id().is_child_of(self.circuit.global_id()) {
+            //             return None;
+            //         }
+            //         let color = if replay_sources.contains_key(&node.local_id()) {
+            //             Some(0xff5555)
+            //         } else if participate_in_backfill.contains(&node.local_id()) {
+            //             Some(0x5555ff)
+            //         } else {
+            //             None
+            //         };
+            //         Some(DotNodeAttributes::new().with_color(color))
+            //     },
+            //     |edge| {
+            //         let style = if edge.is_dependency() {
+            //             Some("dotted".to_string())
+            //         } else {
+            //             None
+            //         };
+            //         let label = if let Some(stream) = &edge.stream {
+            //             Some(format!("consumers: {}", stream.num_consumers()))
+            //         } else {
+            //             None
+            //         };
+            //         Some(
+            //             DotEdgeAttributes::new(edge.stream_id())
+            //                 .with_style(style)
+            //                 .with_label(label),
+            //         )
+            //     },
+            //     "replay.dot",
+            // );
+            // info!("CircuitHandle::restore: replay circuit is written to replay.dot");
+
+            let replay_info = BootstrapInfo {
+                replay_sources: replay_sources.clone(),
+                need_backfill: participate_in_backfill.clone(),
+            };
+
+            self.replay_info = Some(replay_info.clone());
+
+            Ok(Some(replay_info))
+        } else {
+            Ok(None)
+        }
+    }
+
+    // Iterative step of computing the set of nodes that participate in the replay phase.
+    //
+    // Find all input streams of nodes in `participate_in_backfill_new`.
+    //
+    // * For streams that can be replayed from a replay source,:
+    //  - add the replay source to `replay_sources` and `participate_in_backfill`.
+    //  - create replay streams from the replay source to each consumer of the
+    //   original stream.
+    // * For other streams, add their origin nodes to `participate_in_backfill`.
+    //
+    // Return the set of nodes newly added to `participate_in_backfill`.
+    fn compute_replay_nodes_step(
+        &self,
+        replay_sources: &mut BTreeMap<NodeId, StreamId>,
+        need_backfill: &BTreeSet<NodeId>,
+        participate_in_backfill_new: BTreeSet<NodeId>,
+        participate_in_backfill: &mut BTreeSet<NodeId>,
+    ) -> Result<BTreeSet<NodeId>, DbspError> {
+        let mut inputs = BTreeSet::new();
+
+        for node_id in participate_in_backfill_new.iter() {
+            // Compute ancestors of node_id, including:
```

Comment:
"ancestors" is not a great name, it is really sources
ancestors also implies transitivity